File: errorEstimation/errorEstimation.scala
object errorEstimation {
  class Estimate(val x: Double, val e: Double) {
    def + (that: Estimate) =
      new Estimate(this.x + that.x, this.e + that.e)
    def - (that: Estimate) =
      new Estimate(this.x - that.x, this.e + that.e)
    def * (that: Estimate) =
      new Estimate(this.x * that.x,
                   this.x.abs*that.e+that.x.abs*this.e+this.e*that.e)
    def / (that: Estimate) =
      new Estimate(this.x/that.x, 
                   (this.x.abs*that.e+that.x.abs*this.e)/(that.x.abs*(that.x.abs-that.e)))
    def +- (e2: Double) =
      new Estimate(x,e+e2)
    override def toString =
      x + " +- " + e
  }
  implicit def double2estimate(x: Double): Estimate = new Estimate(x,0)
  implicit def int2estimate(x: Int): Estimate = new Estimate(x,0)

  def main(args: Array[String]) = {
    println(({ x: Estimate => x+2*x+3*x*x})(1 +- 0.1))
    // 6.0 +- 0.93
    println(({ x: Estimate => 
               val y = x+x
               y*y + 2 })(1 +- 0.1))
    // 6.0 +- 0.84
    def poly(x: Estimate) = x+2*x+3/(x*x)
    println(poly(3.0 +- 0.1))
    // 9.33333 +- 0.3242352
    println(poly(30271.3 +- 0.0001))
    // 90813.9 +- 0.0003
    println(((x: Estimate) => poly(x*x))(3 +- 1.0))
    // 27.037 +- 20.931
  }
}

File: errorEstimation/errorEstimation.hs
data Estimate a = Estimate a a deriving Eq

instance Show a => Show (Estimate a) where
    show (Estimate v e) = (show v) ++ " +- " ++ (show e)

instance Num a => Num (Estimate a) where
    (Estimate v1 e1) + (Estimate v2 e2) = Estimate (v1+v2) (e1+e2)
    (Estimate v1 e1) - (Estimate v2 e2) = Estimate (v1-v2) (e1+e2)
    (Estimate v1 e1) * (Estimate v2 e2) = Estimate (v1*v2) ((abs v1)*e2+(abs v2)*e1+e1*e2)
    abs (Estimate v e) = Estimate (abs v) e
    signum (Estimate v e) = Estimate (signum v) (signum (v+e) - signum (v-e))
    fromInteger n = Estimate (fromInteger n) 0

instance Fractional a => Fractional (Estimate a) where
    (Estimate v1 e1) / (Estimate v2 e2) = Estimate (v1/v2) ((av1*e2+av2*e1)/(av2*(av2-e2)))
                                          where av1 = abs v1
                                                av2 = abs v2
    fromRational n = Estimate (fromRational n) 0

(+-) v e = Estimate v e

--- (\x -> x+2*x+3*x*x)(1 +- 0.1)
-- | 6.0 +- 0.9300000000000002
--- (\x -> let y = x+x in y*y + 2)(1 +- 0.1)
-- | 6.0 +- 0.8400000000000001
poly x = x+2*x+3/(x*x)
--- poly(3.0 +- 0.1)
-- | 9.333333333333334 +- 0.32423520063567746
--- (\x -> poly (x*x)) (3.0 +- 1.0)
-- | 27.037037037037038 +- 20.93104806934594

